src/libmongoc/src/mongoc/mongoc-database.c

```diff
@@ -996,6 +995,293 @@ mongoc_database_create_collection (mongoc_database_t *database,
+   ecocName = _mongoc_get_encryptedField_state_collection (
+      encryptedFields, name, "ecoc", error);
+   if (!ecocName) {
+      goto fail;
+   }
+
+   ecocCollection = create_collection (database, ecocName, NULL /* opts */, error);
+   if (!ecocCollection) {
+      goto fail;
+   }
+
+   /* Create data collection. */
+   cc_opts = bson_copy (opts);
+   if (!BSON_APPEND_DOCUMENT (cc_opts, "encryptedFields", encryptedFields)) {
+      bson_set_error (error,
+                        MONGOC_ERROR_COMMAND,
+                        MONGOC_ERROR_COMMAND_INVALID_ARG,
+                        "unable to append encryptedFields");
+      goto fail;
+   }
+   dataCollection = create_collection (database, name, cc_opts, error);
+   if (!dataCollection) {
+      goto fail;
+   }
+
+   /* Create index on __safeContent__. */
+   {
+      bson_t *keys = BCON_NEW ("__safeContent__", BCON_INT32 (1));
+      char *index_name;
+      bson_t *create_indexes;
+
+      index_name = mongoc_collection_keys_to_index_string (keys);
+      create_indexes = BCON_NEW ("createIndexes",
+                                 BCON_UTF8 (name),
+                                 "indexes",
+                                 "[",
+                                 "{",
+                                 "key",
+                                 BCON_DOCUMENT (keys),
+                                 "name",
+                                 BCON_UTF8 (index_name),
+                                 "}",
+                                 "]");
+
+      if (!mongoc_database_write_command_with_opts (database,
+                                                    create_indexes,
+                                                    NULL /* opts */,
+                                                    NULL /* reply */,
+                                                    error)) {
+         bson_destroy (create_indexes);
+         bson_free (index_name);
+         bson_destroy (keys);
+         goto fail;
+      }
+      bson_destroy (create_indexes);
+      bson_free (index_name);
+      bson_destroy (keys);
+   }
+
+   ok = true;
+fail:
+   bson_destroy (cc_opts);
+   mongoc_collection_destroy (ecocCollection);
+   bson_free (ecocName);
+   mongoc_collection_destroy (eccCollection);
+   bson_free (eccName);
+   mongoc_collection_destroy (escCollection);
+   bson_free (escName);
+   if (ok) {
+      return dataCollection;
+   } else {
+      mongoc_collection_destroy (dataCollection);
+      return NULL;
+   }
+}
+
+bool
+_mongoc_get_encryptedFields_from_map (mongoc_client_t *client,
+                                      const char *dbName,
+                                      const char *collName,
+                                      bson_t *encryptedFields,
+                                      bson_error_t *error)
+{
+   const bson_t *efMap = client->topology->encrypted_fields_map;
+
+   bson_init (encryptedFields);
+
+   if (bson_empty0 (efMap)) {
+      return true;
```

Comment:
I'd recommend comments on reasons for early returns and `goto fail`s. What conditions are being satisfied or violated? e.g. "Empty efMap will have no encrypted fields", "No efMap entry for this database+collection", or "The efMap entry should always be a document", to help a future reader quickly understand the purpose of a branch or jump.